rust/arrow/src/compute/kernels/length.rs

```diff
@@ -17,52 +17,56 @@
 
 //! Defines kernel for length of a string array
 
-use crate::array::*;
+use crate::datatypes::ToByteSlice;
+use crate::{array::*, buffer::Buffer};
 use crate::{
     datatypes::DataType,
-    datatypes::UInt32Type,
     error::{ArrowError, Result},
 };
 use std::sync::Arc;
 
-/// Returns an array of UInt32 denoting the number of characters in each string in the array.
+fn length_string<OffsetSize>(array: &Array, data_type: DataType) -> Result<ArrayRef>
+where
+    OffsetSize: OffsetSizeTrait,
+{
+    // note: offsets are stored as u8, but they can be interpreted as OffsetSize
+    let offsets = array.data_ref().clone().buffers()[0].clone();
+    // this is a 30% improvement over iterating over u8s and building OffsetSize, which
+    // justifies the usage of `unsafe`.
+    let slice: &[OffsetSize] = unsafe { offsets.typed_data::<OffsetSize>() };
```

Comment:
To support sliced arrays this needs to take the offset of the array into account. The following should work, but a testcase would be nice:

```suggestion
    let slice: &[OffsetSize] = unsafe { offsets.typed_data::<OffsetSize>() }[array.offset()..];
```

Reply:
Great finding, @jhorstmann - fixed and added test.